python/rascaline/rascaline/utils/clebsch_gordan/_utils.py

```diff
@@ -189,82 +186,130 @@ def _parse_bool_iteration_filters(
+    This function returns the computed keys, and a list of tuple of integers, indicating
+    for each entries in the output keys which entry of the two set of input keys should
+    be combined together.
 
-    If `skip_redundant` is True, then keys that represent redundant CG
-    operations are not included in the output keys at each step.
+    The output keys are generated from a full product of ``keys_1`` and ``keys_2``.
+    Then, key are selected according to the angular and/or parity selections specified
+    in ``selected_keys``.
+
+    If ``skip_redundant`` is True, then keys that represent redundant CG operations are
+    not included in the output keys at each step.
     """
-    keys_metadata: List[Tuple[List[LabelsEntry], List[LabelsEntry], Labels]] = []
-    keys_out = keys_1
-    for iteration in range(n_iterations):
-        # Get the keys metadata for the combination of the 2 tensors
-        keys_1_entries, keys_2_entries, keys_out = _precompute_keys_full_product(
-            keys_1=keys_out,
-            keys_2=keys_2,
+
+    # Get the keys metadata for the combination of the 2 tensors
+    output_keys, combinations = _precompute_keys_full_product(keys_1, keys_2)
+
+    if selected_keys is not None:
+        output_keys, combinations = _apply_key_selection(
+            output_keys,
+            combinations,
+            selected_keys,
         )
-        # For TorchScript to determine the type correctly so we can subscript it
-        selected_keys_i = selected_keys[iteration]
-        if selected_keys_i is not None:
-            keys_1_entries, keys_2_entries, keys_out = _apply_key_selection(
-                keys_1_entries,
-                keys_2_entries,
-                keys_out,
-                selected_keys=selected_keys_i,
-            )
 
-        if skip_redundant[iteration]:
-            keys_1_entries, keys_2_entries, keys_out = _remove_redundant_keys(
-                keys_1_entries, keys_2_entries, keys_out
-            )
+    if skip_redundant:
+        output_keys, combinations = _remove_redundant_keys(
+            output_keys,
+            combinations,
+            nu_1=keys_1.column("order_nu")[0],
+            nu_2=keys_2.column("order_nu")[0],
+        )
 
-        # Check that some keys are produced as a result of the combination
-        if len(keys_out) == 0:
-            raise ValueError(
-                f"invalid selections: iteration {iteration + 1} produces no"
-                " valid combinations. Check the `angular_cutoff` and"
-                " `selected_keys` args and try again."
-            )
+    output_keys, combinations = _group_combinations_of_same_blocks(
+        output_keys, combinations
+    )
 
-        keys_metadata.append((keys_1_entries, keys_2_entries, keys_out))
+    return output_keys, combinations
+
+
+def _group_combinations_of_same_blocks(
+    output_keys: Labels,
+    combinations: List[Tuple[int, int]],
+) -> Tuple[Labels, List[Combination]]:
+    """TODO"""
```

Comment:
A stray todo

Reply:
updated, let me know if the docstring is clear

Reply:
looks good, thanks!